### PR TITLE
migrate password resets to redis

### PR DESCRIPTION
**Pull request by kevintsoii**

### CODE REVIEW

File: api/main_endpoints/routes/Auth.js [hunk 1]
```diff
@@ -103,11 +104,7 @@ router.post('/sendPasswordReset', async (req, res) => {
 
     const resetToken = id.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
     try {
-      await PasswordReset.updateOne(
-        { userId: result._id },
-        { $set: { token: resetToken, createdAt: new Date() } },
-        { upsert: true }
-      );
+      await redisClient.set(resetToken, String(result._id), {EX: 60 * 60 * 12}); // 12 hour expiration
```
Comment (evanugarte) [inline api/main_endpoints/routes/Auth.js]: make EX a constant variable in `api/util/constants`

File: api/main_endpoints/util/userHelpers.js [hunk 2]
```diff
@@ -163,12 +163,14 @@ function hashPassword(password) {
...
-      .catch(err => reject(err));
-  });
+async function findPasswordReset(resetToken) {
+  try {
+    const userId = await redisClient.get(resetToken);
+    return userId;
+  } catch (error) {
+    logger.error('Unable to get reset token:', error);
+    throw error;
+  }
```
Comment (evanugarte) [inline api/main_endpoints/util/userHelpers.js]:
```suggestion
    let userId = null;
    try {
    userId = await redisClient.get(resetToken);
  } catch (error) {
    logger.error('Unable to get reset token:', error);
  }
  return userId;
```

File: docker-compose.yml [hunk 3]
```diff
@@ -1,5 +1,11 @@
 version: '3.8'
 services:
+  redis:
+    container_name: sce-redis-dev
+    image: redis:latest
+    restart: always
+    volumes:
+      - sce_redis_data_dev:/data/redis
```
Comment (evanugarte) [inline docker-compose.yml]: make `sce_redis_data` a volume at the bottom of the file
```suggestion
      - sce_redis_data:/data/redis
```

Comment (kevintsoii) [reply docker-compose.yml]: done

File: api/main_endpoints/util/redis-client.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  async del(key) {
+    return this.store.delete(key) ? true : false;
+  }
+}
+
+const redisHost = process.env.REDIS_HOST;
+
+const client = redisHost
+  ? createClient({ url: `redis://${redisHost}:6379` })
+  : new MockRedisClient();
```
Comment (evanugarte) [inline api/main_endpoints/util/redis-client.js]: could we put this in the class instead, and call the class `SceRedisClient`

in the constructor, the class decides whether to use the redis client or the js map

File: api/main_endpoints/util/redis-client.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  async del(key) {
+    return this.store.delete(key) ? true : false;
+  }
+}
+
+const redisHost = process.env.REDIS_HOST;
+
+const client = redisHost
+  ? createClient({ url: `redis://${redisHost}:6379` })
```
Comment (evanugarte) [inline api/main_endpoints/util/redis-client.js]: can we change the environment variable to be the entire url, not just the host? that way we dont need to format the string before calling createClient

the variable could be renamed to `REDIS_HOST_URL`

File: api/util/constants.js [hunk 6]
```diff
@@ -30,9 +30,12 @@ const consoleColors = {
   defaultColor: '\x1b[0m',
 };
 
+const PASSWORD_RESET_EXPIRATION = 60 * 60 * 12;
```
Comment (evanugarte) [inline api/util/constants.js:35]: add a comment above this variable about what this means
```js
// this is set to 12 hours by:
// 60 seconds per minute * 60 minutes per hour * 12 hours
```

Comment (kevintsoii) [reply api/util/constants.js:35]: done

File: api/main_endpoints/util/redis-client.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  async connect() {
+    if (this.useRedis) {
+      try {
+        await this.client.connect();
+      } catch (err) {
+        logger.error('Error connecting to Redis:', err);
+      }
+    }
```
Comment (evanugarte) [inline api/main_endpoints/util/redis-client.js]:
```suggestion
  async connect() {
    // skip connecting to redis if it's not configured
    // when we initialized the class
    if (!this.useRedis) {
        return;
    }
    try {
      await this.client.connect();
    } catch (err) {
      logger.error('Error connecting to Redis:', err);
    }
  }
```

File: api/main_endpoints/util/redis-client.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  constructor() {
+    const redisHostUrl = process.env.REDIS_HOST_URL;
+
+    if (redisHostUrl) {
+      this.client = createClient({ url: redisHostUrl });
+      this.useRedis = true;
+    } else {
+      this.client = new Map();
+      this.useRedis = false;
+    }
```
Comment (evanugarte) [inline api/main_endpoints/util/redis-client.js]:
```suggestion
    const redisHostUrl = process.env.REDIS_HOST_URL;
    
    this.useRedis = !!redisHostUrl;
    this.client = new Map();

    if (redisHostUrl) {
      this.client = createClient({ url: redisHostUrl });
    }
```

File: api/main_endpoints/util/redis-client.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  async set(key, value, options = {}) {
+    return this.client.set(key, value, options);
+  }
+
+  async get(key) {
+    return this.client.get(key) || null;
+  }
+
+  async del(key) {
```
Comment (evanugarte) [inline api/main_endpoints/util/redis-client.js]: call this function `delete`? up to you

**Review by evanugarte — APPROVED** [review]

**Review by pnavab — APPROVED** [review]

**Pull request merged**